File: mac_scripts/functions/brew-desired-update/update.mjs
import { createBrewRunner, ensureTap, loadBrewState, resolveBrewBinary } from "./brew.mjs";
import { loadDesiredDocument } from "./config.mjs";
import { formatNameList } from "./ui.mjs";

const SUBCOMMANDS = new Set(["add", "remove", "list", "ls", "help"]);

/**
 * @param {string[]} argv
 */
export function parseArgv(argv) {
  const exclude = [];
  let i = 0;
  if (argv[0] === "-e" || argv[0] === "--exclude") {
    i = 1;
    while (
      i < argv.length &&
      !argv[i].startsWith("-") &&
      !SUBCOMMANDS.has(argv[i])
    ) {
      exclude.push(argv[i++]);
    }
  }
  const action = argv[i];
  const rest = argv.slice(i + (action ? 1 : 0));
  return { exclude, action, rest };
}

/**
 * @param {string[]} desiredCasks
 * @param {string[]} installedCasks
 */
export function collectCasksToUpgrade(desiredCasks, installedCasks) {
  const has = new Set(installedCasks);
  return desiredCasks.filter((cask) => has.has(cask));
}

/**
 * @param {string[]} casks
 * @param {string[]} exclude
 */
export function filterExcludedCasks(casks, exclude) {
  const excluded = new Set(exclude);
  return casks.filter((cask) => !excluded.has(cask));
}

/**
 * @param {{ exclude?: string[] }} options
 * @param {object} context
 */
export async function runUpdateCommand({ exclude = [] } = {}, context = {}) {
  const {
    env = process.env,
    ui,
    resolveBrew = resolveBrewBinary,
    loadDesired = loadDesiredDocument,
    loadBrewState: loadState = loadBrewState,
    ensureTap: ensureTapFn = ensureTap,
    runBrew: runBrewFn,
  } = context;

  const brewBin = context.brewBin ?? resolveBrew({ env });
  if (!brewBin) {
    ui.error("Homebrew not found.");
    return 1;
  }

  const runner = runBrewFn ?? createBrewRunner({ brewBin, ui });

  const loaded = await loadDesired({ env });
  if (!loaded.ok) {
    ui.error(loaded.error);
    return 1;
  }

  const { document } = loaded;

  ui.title();
  ui.step(
    `Desired lists: ${document.formulas.length} formulae · ${document.taps.length} taps · ${document.casks.length} casks`,
  );
  ui.step("Loading Homebrew state");
  let state = await loadState({ brewBin, runBrew: runner });

  const missingTaps = document.taps.filter((t) => !state.taps.includes(t));
  if (missingTaps.length > 0) {
    ui.active(`Ensuring taps · ${missingTaps.length} missing: ${formatNameList(missingTaps)}`);
    for (const tap of missingTaps) {
      const ok = await ensureTapFn(tap, { brewBin, runBrew: runner });
      if (!ok) {
        ui.error(`Failed to tap ${tap}`);
        return 1;
      }
    }
  } else {
    ui.step("Ensuring taps · all present");
  }

  ui.step("Updating Homebrew (brew update)");
  let result = await runner(["update"]);
  if (result.code !== 0) {
    ui.error("brew update failed");
    return 1;
  }

  ui.active("Upgrading formulae (brew upgrade --formula)");
  result = await runner(["upgrade", "--formula", "-y"]);
  if (result.code !== 0) {
    ui.error("brew upgrade --formula failed");
    return 1;
  }

  ui.step("Repairing taps");
  result = await runner(["tap", "--repair"]);
  if (result.code !== 0) {
    ui.error("brew tap --repair failed");
    return 1;
  }

  ui.step("Reloading Homebrew state");
  state = await loadState({ brewBin, runBrew: runner });

  const eligible = collectCasksToUpgrade(document.casks, state.casks);
  const toUpgrade = filterExcludedCasks(eligible, exclude);
  const eligibleCount = eligible.length;

  if (exclude.length > 0) {
    ui.info(`Excluding ${exclude.length} cask(s): ${formatNameList(exclude)}`);
  }

  if (toUpgrade.length > 0) {
    ui.active(
      `Upgrading casks · ${toUpgrade.length} of ${eligible.length} eligible: ${formatNameList(toUpgrade)}`,
    );
    result = await runner(["upgrade", "--cask", "-y", ...toUpgrade]);
    if (result.code !== 0) {
      ui.error("brew upgrade --cask failed");
      return 1;
    }
  } else if (eligibleCount > 0 && exclude.length > 0) {
    ui.info("All eligible casks were excluded; no cask upgrade.");
  } else {
    ui.info("No casks in list are installed; skipping cask upgrade.");
  }

  ui.step("Cleanup (brew cleanup --prune=1)");
  result = await runner(["cleanup", "--prune=1"]);
  if (result.code !== 0) {
    ui.error("brew cleanup failed");
    return 1;
  }

  return 0;
}

File: mac_scripts/functions/brew-desired-update/update.py
from brew import ensure_tap, load_brew_state
from messages import fmt


def collect_casks_to_upgrade(desired_casks, installed) -> list[str]:
    has = set(installed)
    return [cask for cask in desired_casks if cask in has]


def filter_excluded(casks, exclude) -> list[str]:
    excluded = set(exclude)
    return [cask for cask in casks if cask not in excluded]


def _format_name_list(names, limit: int = 8) -> str:
    if not names:
        return ""
    if len(names) <= limit:
        return " ".join(names)
    shown = " ".join(names[:limit])
    return f"{shown} … (+{len(names) - limit} more)"


def run_update(exclude, document, runner, ui) -> int:
    formulas = document.get("formulas") or []
    taps = document.get("taps") or []
    casks = document.get("casks") or []

    ui.active(
        f"Desired lists: {len(formulas)} formulae · {len(taps)} taps · {len(casks)} casks"
    )
    ui.step("Loading Homebrew state")
    state = load_brew_state(runner, ui)
    ui.active("Loaded Homebrew state")

    missing_taps = [tap for tap in taps if tap not in state["taps"]]
    if missing_taps:
        missing_label = (
            f"Ensuring taps · {len(missing_taps)} missing: {_format_name_list(missing_taps)}"
        )
        ui.step(missing_label)
        for tap in missing_taps:
            if not ensure_tap(tap, runner):
                ui.error(fmt("tap.failed", name=tap))
                return 1
        ui.active(
            f"Ensured taps · {len(missing_taps)} missing: {_format_name_list(missing_taps)}"
        )
    else:
        ui.active("Ensured taps · all present")

    ui.step("Updating Homebrew")
    result = runner.run(["update"], ui)
    if result.returncode != 0:
        ui.error(fmt("brew.update.failed"))
        return 1
    ui.active("Updated Homebrew")

    ui.step("Upgrading formulae")
    result = runner.run(["upgrade", "--formula", "-y"], ui)
    if result.returncode != 0:
        ui.error(fmt("brew.upgrade.formula.failed"))
        return 1
    ui.active("Upgraded formulae")

    ui.step("Repairing taps")
    result = runner.run(["tap", "--repair"], ui)
    if result.returncode != 0:
        ui.error(fmt("brew.tap.repair.failed"))
        return 1
    ui.active("Repaired taps")

    ui.step("Reloading Homebrew state")
    state = load_brew_state(runner, ui)
    ui.active("Reloaded Homebrew state")

    eligible = collect_casks_to_upgrade(casks, state["casks"])
    to_upgrade = filter_excluded(eligible, exclude)
    if exclude:
        ui.info(fmt("exclude.info", n=len(exclude), names=_format_name_list(exclude)))

    if to_upgrade:
        ui.step(f"Upgrading casks · {len(to_upgrade)} of {len(eligible)} eligible")
        result = runner.run(["upgrade", "--cask", "-y", *to_upgrade], ui)
        if result.returncode != 0:
            ui.error(fmt("brew.upgrade.cask.failed"))
            return 1
        ui.active(f"Upgraded casks · {len(to_upgrade)} of {len(eligible)} eligible")
    elif eligible and exclude:
        ui.info(fmt("casks.all_excluded"))
    else:
        ui.info(fmt("casks.none_installed"))

    ui.step("Cleaning up Homebrew")
    result = runner.run(["cleanup", "--prune=1"], ui)
    if result.returncode != 0:
        ui.error(fmt("brew.cleanup.failed"))
        return 1
    ui.active("Cleaned up Homebrew")
    return 0
